File: src/main/java/com/xiangqi/rules/MoveValidator.java
package com.xiangqi.rules;

import com.xiangqi.model.*;
import com.xiangqi.game.ChessBoard;
import org.springframework.stereotype.Component;

@Component
public class MoveValidator {

    public boolean isValidMove(ChessBoard board, Move move, PieceColor currentTurn) {
        Position from = move.getFrom();
        Position to = move.getTo();

        Piece piece = board.getPiece(from);
        if (piece == null || piece.getColor() != currentTurn) return false;
        if (!to.isValid()) return false;

        Piece targetPiece = board.getPiece(to);
        if (targetPiece != null && targetPiece.getColor() == currentTurn) return false;

        switch (piece.getType()) {
            case GENERAL: return validateGeneralMove(board, from, to, piece.getColor());
            case ADVISOR: return validateAdvisorMove(board, from, to, piece.getColor());
            case ELEPHANT: return validateElephantMove(board, from, to, piece.getColor());
            case HORSE: return validateHorseMove(board, from, to);
            case CHARIOT: return validateChariotMove(board, from, to);
            case CANNON: return validateCannonMove(board, from, to);
            case SOLDIER: return validateSoldierMove(from, to, piece.getColor());
            default: return false;
        }
    }

    private boolean validateGeneralMove(ChessBoard board, Position from, Position to, PieceColor color) {
        if (!to.isInPalace(color)) return false;
        int rowDiff = Math.abs(to.getRow() - from.getRow());
        int colDiff = Math.abs(to.getCol() - from.getCol());
        return (rowDiff + colDiff == 1);
    }

    private boolean validateAdvisorMove(ChessBoard board, Position from, Position to, PieceColor color) {
        if (!to.isInPalace(color)) return false;
        int rowDiff = Math.abs(to.getRow() - from.getRow());
        int colDiff = Math.abs(to.getCol() - from.getCol());
        return (rowDiff == 1 && colDiff == 1);
    }

    private boolean validateElephantMove(ChessBoard board, Position from, Position to, PieceColor color) {
        if (color == PieceColor.RED && to.getRow() < 5) return false;
        if (color == PieceColor.BLACK && to.getRow() > 4) return false;

        int rowDiff = to.getRow() - from.getRow();
        int colDiff = to.getCol() - from.getCol();

        if (Math.abs(rowDiff) != 2 || Math.abs(colDiff) != 2) return false;

        int eyeRow = from.getRow() + rowDiff / 2;
        int eyeCol = from.getCol() + colDiff / 2;
        Position eyePos = new Position(eyeRow, eyeCol);

        return board.getPiece(eyePos) == null;
    }

    private boolean validateHorseMove(ChessBoard board, Position from, Position to) {
        int rowDiff = Math.abs(to.getRow() - from.getRow());
        int colDiff = Math.abs(to.getCol() - from.getCol());

        if (!((rowDiff == 2 && colDiff == 1) || (rowDiff == 1 && colDiff == 2))) return false;

        int legRow, legCol;
        if (rowDiff == 2) {
            legRow = from.getRow() + (to.getRow() - from.getRow()) / 2;
            legCol = from.getCol();
        } else {
            legRow = from.getRow();
            legCol = from.getCol() + (to.getCol() - from.getCol()) / 2;
        }

        Position legPos = new Position(legRow, legCol);
        return board.getPiece(legPos) == null;
    }

    private boolean validateChariotMove(ChessBoard board, Position from, Position to) {
        if (from.getRow() != to.getRow() && from.getCol() != to.getCol()) return false;
        return countPiecesBetween(board, from, to) == 0;
    }

    private boolean validateCannonMove(ChessBoard board, Position from, Position to) {
        if (from.getRow() != to.getRow() && from.getCol() != to.getCol()) return false;

        int piecesBetween = countPiecesBetween(board, from, to);
        Piece targetPiece = board.getPiece(to);

        if (targetPiece == null) {
            return piecesBetween == 0;
        } else {
            return piecesBetween == 1;
        }
    }

    private boolean validateSoldierMove(Position from, Position to, PieceColor color) {
        int rowDiff = to.getRow() - from.getRow();
        int colDiff = Math.abs(to.getCol() - from.getCol());

        if (color == PieceColor.RED) {
            if (from.getRow() > 4) {
                return (rowDiff == -1 && colDiff == 0);
            } else {
                return ((rowDiff == -1 && colDiff == 0) || (rowDiff == 0 && colDiff == 1));
            }
        } else {
            if (from.getRow() < 5) {
                return (rowDiff == 1 && colDiff == 0);
            } else {
                return ((rowDiff == 1 && colDiff == 0) || (rowDiff == 0 && colDiff == 1));
            }
        }
    }

    private int countPiecesBetween(ChessBoard board, Position from, Position to) {
        int count = 0;

        if (from.getRow() == to.getRow()) {
            int minCol = Math.min(from.getCol(), to.getCol());
            int maxCol = Math.max(from.getCol(), to.getCol());
            for (int col = minCol + 1; col < maxCol; col++) {
                if (board.getPiece(new Position(from.getRow(), col)) != null) count++;
            }
        } else {
            int minRow = Math.min(from.getRow(), to.getRow());
            int maxRow = Math.max(from.getRow(), to.getRow());
            for (int row = minRow + 1; row < maxRow; row++) {
                if (board.getPiece(new Position(row, from.getCol())) != null) count++;
            }
        }

        return count;
    }
}

File: src/game/MoveValidator.ts
import { ChessBoard } from './ChessBoard';
import { Piece, PieceColor, Position, Move } from './types';

export class MoveValidator {
  isValidMove(board: ChessBoard, move: Move, currentTurn: PieceColor): boolean {
    const from = move.from;
    const to = move.to;

    const piece = board.getPiece(from);
    if (piece === null || piece.color !== currentTurn) return false;
    if (!board.isValidPosition(to)) return false;

    const targetPiece = board.getPiece(to);
    if (targetPiece !== null && targetPiece.color === currentTurn) return false;

    switch (piece.type) {
      case 'GENERAL': return this.validateGeneralMove(board, from, to, piece.color);
      case 'ADVISOR': return this.validateAdvisorMove(board, from, to, piece.color);
      case 'ELEPHANT': return this.validateElephantMove(board, from, to, piece.color);
      case 'HORSE': return this.validateHorseMove(board, from, to);
      case 'CHARIOT': return this.validateChariotMove(board, from, to);
      case 'CANNON': return this.validateCannonMove(board, from, to);
      case 'SOLDIER': return this.validateSoldierMove(from, to, piece.color);
      default: return false;
    }
  }

  private validateGeneralMove(board: ChessBoard, from: Position, to: Position, color: PieceColor): boolean {
    if (!board.isInPalace(to, color)) return false;
    const rowDiff = Math.abs(to.row - from.row);
    const colDiff = Math.abs(to.col - from.col);
    return (rowDiff + colDiff === 1);
  }

  private validateAdvisorMove(board: ChessBoard, from: Position, to: Position, color: PieceColor): boolean {
    if (!board.isInPalace(to, color)) return false;
    const rowDiff = Math.abs(to.row - from.row);
    const colDiff = Math.abs(to.col - from.col);
    return (rowDiff === 1 && colDiff === 1);
  }

  private validateElephantMove(board: ChessBoard, from: Position, to: Position, color: PieceColor): boolean {
    // Cannot cross river
    if (color === PieceColor.RED && to.row < 5) return false;
    if (color === PieceColor.BLACK && to.row > 4) return false;

    const rowDiff = to.row - from.row;
    const colDiff = to.col - from.col;

    if (Math.abs(rowDiff) !== 2 || Math.abs(colDiff) !== 2) return false;

    // Check blocking piece at elephant eye
    const eyeRow = from.row + rowDiff / 2;
    const eyeCol = from.col + colDiff / 2;

    return board.getPiece({ row: eyeRow, col: eyeCol }) === null;
  }

  private validateHorseMove(board: ChessBoard, from: Position, to: Position): boolean {
    const rowDiff = Math.abs(to.row - from.row);
    const colDiff = Math.abs(to.col - from.col);

    if (!((rowDiff === 2 && colDiff === 1) || (rowDiff === 1 && colDiff === 2))) return false;

    // Check blocking piece at horse leg
    let legRow: number, legCol: number;
    if (rowDiff === 2) {
      legRow = from.row + (to.row - from.row) / 2;
      legCol = from.col;
    } else {
      legRow = from.row;
      legCol = from.col + (to.col - from.col) / 2;
    }

    return board.getPiece({ row: legRow, col: legCol }) === null;
  }

  private validateChariotMove(board: ChessBoard, from: Position, to: Position): boolean {
    if (from.row !== to.row && from.col !== to.col) return false;
    return this.countPiecesBetween(board, from, to) === 0;
  }

  private validateCannonMove(board: ChessBoard, from: Position, to: Position): boolean {
    if (from.row !== to.row && from.col !== to.col) return false;

    const piecesBetween = this.countPiecesBetween(board, from, to);
    const targetPiece = board.getPiece(to);

    if (targetPiece === null) {
      // Moving without capture - must have no pieces in between
      return piecesBetween === 0;
    } else {
      // Capturing - must have exactly one piece in between (cannon platform)
      return piecesBetween === 1;
    }
  }

  private validateSoldierMove(from: Position, to: Position, color: PieceColor): boolean {
    const rowDiff = to.row - from.row;
    const colDiff = Math.abs(to.col - from.col);

    if (color === PieceColor.RED) {
      if (from.row > 4) {
        // Before crossing river - can only move forward
        return (rowDiff === -1 && colDiff === 0);
      } else {
        // After crossing river - can move forward or sideways
        return ((rowDiff === -1 && colDiff === 0) || (rowDiff === 0 && colDiff === 1));
      }
    } else {
      if (from.row < 5) {
        return (rowDiff === 1 && colDiff === 0);
      } else {
        return ((rowDiff === 1 && colDiff === 0) || (rowDiff === 0 && colDiff === 1));
      }
    }
  }

  private countPiecesBetween(board: ChessBoard, from: Position, to: Position): number {
    let count = 0;

    if (from.row === to.row) {
      const minCol = Math.min(from.col, to.col);
      const maxCol = Math.max(from.col, to.col);
      for (let col = minCol + 1; col < maxCol; col++) {
        if (board.getPiece({ row: from.row, col }) !== null) count++;
      }
    } else {
      const minRow = Math.min(from.row, to.row);
      const maxRow = Math.max(from.row, to.row);
      for (let row = minRow + 1; row < maxRow; row++) {
        if (board.getPiece({ row, col: from.col }) !== null) count++;
      }
    }

    return count;
  }
}